<img align="left" src="https://lever-client-logos.s3.amazonaws.com/864372b1-534c-480e-acd5-9711f850815c-1524247202159.png" width=200>
<br></br>
<br></br>

## *Data Science Unit 1 Sprint 2 Assignment 1*

# Apply the t-test to real data

## Practice 1-Sample T-tests

One Sample t-tests determine whether or not a sample mean is statistically different from some known (or hypothesized) population mean. 

### 1) Load the Data
- Use the [automobile dataset](https://archive.ics.uci.edu/ml/datasets/Automobile)
- Fix the column headers
- Make sure NaNs are used to indicate missing values

Feel free to add code cells and text cells as needed throughout the assignment.

In [ ]:
### YOUR WORK HERE
import pandas as pd
import numpy as np

#Load the data
url = 'https://archive.ics.uci.edu/ml/machine-learning-databases/autos/imports-85.data'
column_headers = ['symboling', 'normalized-losses', 'make', 'fuel-type', 'aspiration', 'num-of-doors', 'body-style', 'drive-wheels', 'engine-locaton', 'wheel-base', 'length', 'width', 'height', 'curb-weight', 'enging-type', 'num-of-cylinders', 'engine-size', 'fuel-system', 'bore', 'stroke', 'compression-ratio', 'horsepower', 'peak-rpm', 'city-mpg', 'highway-mpg', 'price']
df = pd.read_csv(url, names=column_headers)

df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3                 ?  alfa-romero  ...       21          27  13495
1          3                 ?  alfa-romero  ...       21          27  16500
2          1                 ?  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
#Check for missing values
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-locaton        0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
enging-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

In [ ]:
#Replace the missing values '?' with NaNs
df = df.replace({'?':np.NaN})
df.head()

symboling normalized-losses         make  ... city-mpg highway-mpg  price
0          3               NaN  alfa-romero  ...       21          27  13495
1          3               NaN  alfa-romero  ...       21          27  16500
2          1               NaN  alfa-romero  ...       19          26  16500
3          2               164         audi  ...       24          30  13950
4          2               164         audi  ...       18          22  17450

[5 rows x 26 columns]

In [ ]:
df.isnull().sum()

symboling             0
normalized-losses    41
make                  0
fuel-type             0
aspiration            0
num-of-doors          2
body-style            0
drive-wheels          0
engine-locaton        0
wheel-base            0
length                0
width                 0
height                0
curb-weight           0
enging-type           0
num-of-cylinders      0
engine-size           0
fuel-system           0
bore                  4
stroke                4
compression-ratio     0
horsepower            2
peak-rpm              2
city-mpg              0
highway-mpg           0
price                 4
dtype: int64

### 2) Pretend that this dataset represents the cars at a used car lot in your local town. 

- Use df.sample() to pick a random sample of 10 cars. Note that because this sample is random we are going to set the `random_state` so that all of us in the class get the same random sample. Please set your random state to `30` when using `df.sample()`

Is your sample reflective of the population value in regards to highway-mpg? Find the mean for `highway-mpg` for the entire dataset and compare your sample mean estimate to that value. Can you say that they are different? 

In [ ]:
#Extract the sample
my_sample = df.sample(10, random_state=30)

my_sample.head()

symboling normalized-losses           make  ... city-mpg highway-mpg  price
121          1               154       plymouth  ...       31          38   6692
176         -1                65         toyota  ...       27          32  10898
24           1               148          dodge  ...       31          38   6229
88          -1               137     mitsubishi  ...       23          30   9279
72           3               142  mercedes-benz  ...       16          18  35056

[5 rows x 26 columns]

In [ ]:
### YOUR WORK HERE
#Determine the mean for the entire sample
print(df['highway-mpg'].mean())
print(my_sample['highway-mpg'].mean())

30.75121951219512
32.8


###Means for my sample and the population
Mean for the population is 30.75

Mean for my sample is 32.8

These means are approximately 6% in difference, so I would say that they are not the same. 

The salesman says the cars he sells typically have a fuel efficiency of about 35 miles per gallon on the highway. You want to verify his claim but can only test 10 cars. Using your sample of 10, test his claim and report your results.

**Null Hypothesis**: Cars typically sold at this lot have a fuel efficency of about 35mpg on the highway

**Alternative Hypothesis**: Cars typically sold at this lot do not have a fuel efficiency of about 35mpg on the highway.

In [ ]:
### YOUR WORK HERE
from scipy import stats

stats.ttest_1samp(my_sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-0.9570244044334747, pvalue=0.36355472283248624)

###Conclusion

With a t-stat of -0.96 and a pvalue of 0.36, I accept the null hypothesis that the cars typically sold have a fuel efficienty of about 35 mpg on the highway.  

Imagine that you now have the capacity to test 100 cars on the lot. Using the same random state of `30`, what would you conclude when sampling 100 cars?

In [ ]:
### YOUR WORK HERE
my_larger_sample = df.sample(100, random_state=30)

stats.ttest_1samp(my_larger_sample['highway-mpg'], 35)

Ttest_1sampResult(statistic=-6.5185242467032305, pvalue=3.003137417468754e-09)

### Conclusion

With a t-stat of 6.5 and a pvalue below .05, I reject the null hypothesis that the average highway mpg is about 35 mpg.

Why might these two t-tests using the same dataset lead to different conclusions about the salesman's claim?

###Answer:
A larger dataset is more representative of the larger population than a smaller dataset

## Practice 2-Sample T-tests

Two Sample t-tests determine whether or not two sample means are statistically different from each other. 

This portion of your assignment requires you to determine which issues have "statistically significant" differences between political parties in this [1980s congressional voting data](https://archive.ics.uci.edu/ml/datasets/Congressional+Voting+Records). 


### 1) Load the data

The data consists of 435 instances (one for each congressperson), a class (democrat or republican), and 16 binary attributes (yes or no for voting for or against certain issues). Be aware - there are missing values!

- Read the dataset in from UCI, you'll need to provide a list of column headers
- Encode "yes" votes as 1 and "no" votes as 0. (You can use `df.replace()` to swap out these values)
- Use dataframe filtering to split the dataframe into two new dataframes based on party. Hold all republicans in one dataframe and all democrats in the other. These will be our two different "samples."



In [ ]:
### YOUR WORK HERE
# Grab the file from UCI:
!wget https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data

--2020-06-08 21:25:57--  https://archive.ics.uci.edu/ml/machine-learning-databases/voting-records/house-votes-84.data
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 18171 (18K) [application/x-httpd-php]
Saving to: ‘house-votes-84.data’

house-votes-84.data 100%[===================>]  17.75K  --.-KB/s    in 0.06s   

2020-06-08 21:25:58 (287 KB/s) - ‘house-votes-84.data’ saved [18171/18171]



In [ ]:
#Create the column headers
column_headers = ['party', 'handicapped-infants', 'water-project', 'budget',
                  'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
                  'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration',
                  'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free', 'south-africa']

df = pd.read_csv('house-votes-84.data', names=column_headers)

print(df.shape)
df.head()

(435, 17)


party handicapped-infants water-project  ... crime duty-free south-africa
0  republican                   n             y  ...     y         n            y
1  republican                   n             y  ...     y         n            ?
2    democrat                   ?             y  ...     y         n            n
3    democrat                   n             y  ...     n         n            y
4    democrat                   y             y  ...     y         y            y

[5 rows x 17 columns]

In [ ]:
#Replace the yes and no's with 1's and 0's
df = df.replace({'y':1, 'n':0, '?':np.NaN})

df.head()

party  handicapped-infants  ...  duty-free  south-africa
0  republican                  0.0  ...        0.0           1.0
1  republican                  0.0  ...        0.0           NaN
2    democrat                  NaN  ...        0.0           0.0
3    democrat                  0.0  ...        0.0           1.0
4    democrat                  1.0  ...        1.0           1.0

[5 rows x 17 columns]

In [ ]:
#Split the df into two, one for dems, a second for repubs

dems = df[df['party'] == 'democrat']
reps = df[df['party'] == 'republican']

print(df.shape)
print(dems.shape)
print(reps.shape)

(435, 17)
(267, 17)
(168, 17)


In [ ]:
dems.head()

party  handicapped-infants  water-project  ...  crime  duty-free  south-africa
2  democrat                  NaN            1.0  ...    1.0        0.0           0.0
3  democrat                  0.0            1.0  ...    0.0        0.0           1.0
4  democrat                  1.0            1.0  ...    1.0        1.0           1.0
5  democrat                  0.0            1.0  ...    1.0        1.0           1.0
6  democrat                  0.0            1.0  ...    1.0        1.0           1.0

[5 rows x 17 columns]

In [ ]:
reps.head()

party  handicapped-infants  ...  duty-free  south-africa
0   republican                  0.0  ...        0.0           1.0
1   republican                  0.0  ...        0.0           NaN
7   republican                  0.0  ...        NaN           1.0
8   republican                  0.0  ...        0.0           1.0
10  republican                  0.0  ...        0.0           0.0

[5 rows x 17 columns]

### 2) Perform two sample T-tests on different issues and report the results.

- Find an issue that democrats support more than republicans with p < 0.01 (significant at the 99% level).
- Find an issue that republicans support more than democrats with p < 0.01 (significant at the 99% level).
- Find an issue where the difference between republicans and democrats has p > 0.1 (Not significant at the 90% level - i.e. there may not be much of a difference the two sample means)

Please for each test that you run state your null and alternative hypothesis and then write a conclusion reflecting on the null and alternative hypothesis.

Remember, that two-sample t-tests will only tell us if the two groups are *different* from one another. We'll have to look at their sample means directly or use the sign on the t-statistic to know which of the two sample means is larger. 

In [ ]:
### YOUR WORK HERE
#Drop the party column
dems = dems.drop('party', axis=1)
reps = reps.drop('party', axis=1)

#Print each sample's shape to verify the column was dropped (should only be 16 columns afterwards)
print(dems.shape)
print(reps.shape)

(267, 16)
(168, 16)


In [ ]:
#Import stats again because google colab is broken
from scipy import stats

In [ ]:
#Now work through each column running 2-sample ttests on each issue
issues = ['handicapped-infants', 'water-project', 'budget',
                  'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
                  'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration',
                  'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free', 'south-africa']

for i in issues:
  results = str(stats.ttest_ind(reps[i], dems[i], nan_policy='omit'))
  print('The stats for ' + i + ' are:  ' + results)



The stats for handicapped-infants are:  Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)
The stats for water-project are:  Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)
The stats for budget are:  Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)
The stats for physician-fee-freeze are:  Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)
The stats for el-salvador-aid are:  Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)
The stats for religious-groups are:  Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)
The stats for anti-satellite-ban are:  Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)
The stats for aid-to-contras are:  Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)
The stats for mx-missile are:  Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-4

###Conclusions

***Null Hypothesis***: Both parties support the budget bill equally.

***Alternative Hypothesis***: Both parties do not support the budget bill equally.

***Conclusions***: Based on a t-stat of -23.21, and a p-value of 2.07e-77, both parties do not support the budget bill equally.  Democrats support the bill more than Republicans
___
***Null Hypothesis***:  Both parties support the El Savador bill equally.

***Alternative Hypothesis***: Both parties do not support the El Savador Bill equally.

***Conclusion***: Based on a t-stat of 21.14, and a p-value of 5.60e-68, both parties do not support the bill equally.  Republicans support the bill more than democrats
___
***Null Hypothesis***: Both parties support the water bill equally.

***Alternative Hypothesis***: Both parties do not support the water bill equally.  

***Conclusion***: With a t-stat of .089 and a p-value of 0.93, both parties support the water bill equally.  

## Stretch Goals:

### 1) Use functions and some iterator (for loop, .apply(), list comprehension, etc.) to perform two sample t-tests on every issue in the dataset in an automated fashion.

In [ ]:
### YOUR WORK HERE

#Now work through each column running 2-sample ttests on each issue
issues = ['handicapped-infants', 'water-project', 'budget',
                  'physician-fee-freeze', 'el-salvador-aid', 'religious-groups', 
                  'anti-satellite-ban', 'aid-to-contras', 'mx-missile', 'immigration',
                  'synfuels', 'education', 'right-to-sue', 'crime', 'duty-free', 'south-africa']

for i in issues:
  results = str(stats.ttest_ind(reps[i], dems[i], nan_policy='omit'))
  print('The stats for ' + i + ' are:  ' + results)



The stats for handicapped-infants are:  Ttest_indResult(statistic=-9.205264294809222, pvalue=1.613440327937243e-18)
The stats for water-project are:  Ttest_indResult(statistic=0.08896538137868286, pvalue=0.9291556823993485)
The stats for budget are:  Ttest_indResult(statistic=-23.21277691701378, pvalue=2.0703402795404463e-77)
The stats for physician-fee-freeze are:  Ttest_indResult(statistic=49.36708157301406, pvalue=1.994262314074344e-177)
The stats for el-salvador-aid are:  Ttest_indResult(statistic=21.13669261173219, pvalue=5.600520111729011e-68)
The stats for religious-groups are:  Ttest_indResult(statistic=9.737575825219457, pvalue=2.3936722520597287e-20)
The stats for anti-satellite-ban are:  Ttest_indResult(statistic=-12.526187929077842, pvalue=8.521033017443867e-31)
The stats for aid-to-contras are:  Ttest_indResult(statistic=-18.052093200819733, pvalue=2.82471841372357e-54)
The stats for mx-missile are:  Ttest_indResult(statistic=-16.437503268542994, pvalue=5.03079265310811e-4



### 2) Work on Performing a T-test without using Scipy in order to get "under the hood" and learn more thoroughly about this topic.
### Start with a 1-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://blog.minitab.com/hs-fs/hubfs/Imported_Blog_Media/701f9c0efa98a38fb397f3c3ec459b66.png?width=247&height=172&name=701f9c0efa98a38fb397f3c3ec459b66.png) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

### Be sure to check your work using Scipy!





In [ ]:
### YOUR WORK HERE



### 3) Then try a 2-sample t-test
 - Establish the conditions for your test 
 - [Calculate the T Statistic](https://lh3.googleusercontent.com/proxy/rJJ5ZOL9ZDvKOOeBihXoZDgfk7uv1YsRzSQ1Tc10RX-r2HrRpRLVqlE9CWX23csYQXcTniFwlBg3H-qR8MKJPBGnjwndqlhDX3JxoDE5Yg) (You'll need to omit NaN values from your sample).
 - Translate that t-statistic into a P-value. You can use a [table](https://www.google.com/search?q=t+statistic+table) or the [University of Iowa Applet](https://homepage.divms.uiowa.edu/~mbognar/applets/t.html)

 ### Be sure to check your work using Scipy!

In [ ]:
### YOUR WORK HERE